teammapper-backend/src/map/services/maps.service.ts

```diff
@@ -54,45 +54,56 @@ export class MapsService {
     )
   }
 
-  async addNode(mapId: string, clientNode: IMmpClientNode): Promise<MmpNode> {
+  async addNode(mapId: string, clientNode: IMmpClientNode | MmpNode): Promise<MmpNode> {
     // detached nodes are not allowed to have a parent
-    if (clientNode.detached && clientNode.parent) return Promise.reject()
+    const hasParentNode = 'parent' in clientNode ? clientNode.parent : clientNode.nodeParentId
+    if (clientNode.detached && hasParentNode) return Promise.reject()
     if (!mapId || !clientNode) return Promise.reject()
 
     const existingNode = await this.nodesRepository.findOne({
       where: { id: clientNode.id, nodeMapId: mapId },
     })
     if (existingNode) return existingNode
 
-    const newNode = this.nodesRepository.create({
-      ...mapClientNodeToMmpNode(clientNode, mapId),
-      nodeMapId: mapId,
-    })
+    let newNode;
+    // Type guard: IMmpClientNode
+    if ('parent' in clientNode) {
+      newNode = this.nodesRepository.create({
+        ...mapClientNodeToMmpNode(clientNode, mapId),
```

Comment:
I would propose to call this method (the mapping method) before calling "addNode" so we could use just one type (MmpNode) instead of the union type. it would be much more straightforward than making all these case switches.

Reply:
Good point, would simplify the validatesNodeParentForNode() method as well, but we'd have to move the case switches to addNodes()

Reply:
well I would do the mapping then even before addNodes - so we also just have one type here?

Reply:
The problem with that is that addNodes is called both from backend (with MmpNode) and extensively from the frontend (through websocket), so we'd end up doing the mapping in a bunch of different places instead.

Reply:
> The problem with that is that addNodes is called both from backend (with MmpNode) and extensively from the frontend (through websocket), so we'd end up doing the mapping in a bunch of different places instead.

then I would suggest to add a little helper method like `addNodesFromClient` which itself maps the data and then calls `addNodes`?

Reply:
👍